accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+// Can a JSON string be deserialized into a JsonNumber? If so, "JsonNumber" no longer means a "JSON number".
+// When deserializing a string there is no known CLR type, so the internal value would need be based on a string\byte[].
+// So a "JsonString" node would normally be created, not "JsonNumber" and thus throw here:
+number = JsonSerializer.Deserialize<JsonNumber>(json); // Throws
+
+// Even if the above could be made to work, it wouldn't work with a JsonObject node.
+// A JsonObject would create "JsonString" nodes for "NaN" JSON strings since it has
+// no way of knowing these should map to a JsonNumber (or a CLR number).
+```
+
+Thus the DOM is based on CLR types (not JSON tokens) in edit mode and during serialization. The JSON "value" types (number, string, boolean) are represented as a common `JsonValue` sealed class, and not separate `JsonNumber`, `JsonString` and `JsonBoolean` classes as shown in the example above.
+
+During serialization, the DOM is written using the serializer and uses any custom converters that may be configured. In all but idiocentric scenarios (e.g. a custom converter can do anything), the behavior is natural and JSON is round-tripped as expected.
+
+In the example, a JSON "NaN" string is deserialized into a `JsonValue` and when `GetValue<double>` is called, a `double` is returned, assuming the appropriate `JsonSerializerOptions` are configured for quoted numbers. When the node is serialized, it produces a JSON string (again, if configured). The consumer of the DOM is not necessarily aware that `double` values assigned to the node may be serialized to JSON as either a string or a JSON number, and after deserialization is not aware what the internal state is - it could be a `byte[]`, a `JsonElement` or a `string` depending on implementation and\or the JSON token (string or number).
+
+## Primary API
+These types live in STJ.dll.
+```cs
+namespace System.Text.Json
+// Also consider existing S.T.J.Serialization namespace instead of S.T.J.
+// since it is now coupled to the serializer rather than stand-alone.
+{
+    public abstract class JsonNode
+    {
+        public JsonNode(JsonSerializerOptions options = null);
+
+        public JsonSerializerOptions Options { get; }
+
+        // Return the internal value or convert to T as necessary.
+        public abstract T GetValue<T>();
+
+        // The token type from deserialization; otherwise JsonValueKind.Unspecified.
+        public JsonValueKind ValueKind { get; }
+    }
+
+    public sealed class JsonArray : JsonNode, IList<object>
+    {
+        public JsonArray(JsonSerializerOptions options = null);
+        public JsonArray(JsonElement value, JsonSerializerOptions options = null);
+
+        // Can be used to return:
+        // - Any compatible collection type
+        // - JsonElement
+        // - string or byte[] (as JSON)
+        public override T GetValue<T>();
+    }
+
+    public sealed class JsonObject : JsonNode, IDictionary<string, object>
```

Comment:
Why not ` IDictionary<string, JsonNode>`

Reply:
Currently the prototype has `IDictionary<string, JsonNode>`; I'll update the API doc to match.

The original thinking was that someone may want to do this:
```cs
JsonObject jObject = ...
jObject["MyString"] = "hello"; // "hello" is a raw string, not a JsonObject (and no implicit cast from string->JsonObject)
```
and if we have `IDictionary<string, JsonNode>` then you'd have to do this:
```cs
JsonObject jObject = ...
jObject["MyString"] = new JsonValue("hello");
```

Since then, however, indexers were added to `JsonNode` that must return `JsonNode`, not object, in order to get the chaining of arrays and properties so it makes sense to use `IDictionary<string, JsonNode>` and `IList<JsonNode>` for arrays.